[P036] Support for 128x32 display

uwekaditz opened the pull request:
NEW: Support for 128x32 displays (see  https://www.letscontrolit.com/forum/viewtopic.php?p=39840#p39840)
NEW: Option to hide the header
CHG: Calculate font setting, if necessary reduce lines per page (fonts are not longer a fixed setting)
CHG: Calculate font setting for splash screen
CHG: Reduce espeasy_logo to 32x32 to fit all displays

uwekaditz committed:
[P036] Support for 128x32 display

NEW: Support for 128x32 displays (see https://www.letscontrolit.com/forum/viewtopic.php?p=39840#p39840)
NEW: Option to hide the header
CHG: Calculate font setting, if necessary reduce lines per page (fonts are not longer a fixed setting)
CHG: Reduce espeasy_logo to 32x32 to fit all displays
CHG: Calculate font setting for splash screen

Micha-he commented:
Are the changes in the files '_P092_DLbus.ino, P092_data_struct.cpp, P092_data_struct.h' needed for this (P036-) PR ?

uwekaditz committed:
Files for [P092] replaced with original ones from mega

Files for [P092] replaced with original ones from mega to avoid miving with another pull request (##3293 Exception solved)

uwekaditz commented:
No, the files '_P092_DLbus.ino, P092_data_struct.cpp, P092_data_struct.h' are used for another PR #3293 
I will replace these files with the original ones from mega.

By the way, the new features for the P036 plug in are already tested.
They work for the 128x64, 64x48 and 128x32 displays using the SSD1306 chip.

TD-er commented:
Did you also test with those 1.3" displays using the SH1106 chip?

uwekaditz commented:
I'm sorry, I don't have displays with the SH1106 chip. So I couldn't test them.

uwekaditz commented:
According to the datasheet of the SH1106 chip the setting for the 0xDA register (Common pads hardware configuration mode set) is only 0x12 or 0x02 as shown in the picture.
![SH1106 setting for 0xDA](https://user-images.githubusercontent.com/55511593/95736267-4a803800-0c86-11eb-9391-4247296b0760.PNG)

0x12 was the standard used so far.
I added the 0x02 setting, this works for the SSD1306 chip, so it should also work for the SH1106 chips.

TD-er commented:
Well not many people will switch addresses on those OLED displays as those are sometimes really hard to solder without breaking the board.
Also ESP8266 doesn't have enough resources to use multiple instances of this plugin very well (I tried using the I2C multiplexer)
I do have one of those 1.3" units here on my desk, so I will test it also.

Micha-he commented:
@TD-er > Well not many people will switch addresses on those OLED displays as those are sometimes really hard to solder without breaking the board.

Maybe you are wrong ?! The different value of 0x02 instead of 0x12 is a diffent hardware-communication-pin setup of the register 0xDA and not the I2C-adress. The Display with a height of 32px need here a value 0x02.

Or I've misunderstood you both now...

TD-er commented:
Ah I guess you're right. I thought you were talking about I2C addresses. My bad.

Micha-he commented:
Checked with my 128x32 display. Works great !

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
[P016] Improvement: a received IR code can be saved and a command can be assigned to it

uwekaditz committed:
Undo changes for PR #3323 in this branch OLED_P036

TD-er commented:
Why are P016 plugin commits included in this PR?

uwekaditz commented:
I'm sorry, it was a mistake with the P016 files.
I have already removed them as you see in the changed files.

uwekaditz committed:
ressouce-saving string calculation

CHG: ressouce-saving string calculation (see PR #3323)

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -734,7 +740,10 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
 
           if (PixLength > 255) {
-            addHtmlError(String(F("Pixel length of ")) + String(PixLength) + String(F(" too long for line! Max. 255 pix!")));
+            String _error = F("Pixel length of ");
```
Just curious, why using the _ prefix?
Last week I mentioned to @tonhuisman about using the _ prefix for member variables, so if you use them for temporary variables then I guess it could become confusing.

uwekaditz replied:
Change in the next commit.

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -761,22 +770,31 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
-          log += String(F(" Length:")) + String(String(P036_data->DisplayLinesV1[LineNo - 1].Content).length());
-          log += String(F(" Pix: ")) + String(P036_data->display->getStringWidth(String(P036_data->DisplayLinesV1[LineNo - 1].Content)));
-          log += String(F(" Reserved:")) + String(P036_data->DisplayLinesV1[LineNo - 1].reserved);
+          String log = F("[P36] Line: ");
```
These strings for logging can become quite long, and also use quite a lot of operator+
so please try to estimate the final length and use `reserve` on the string before adding content to it.

uwekaditz replied:
Change in the next commit.

uwekaditz committed:
not using the _ prefix anymore, use reserve() for long log strings

uwekaditz committed:
Update P036_data_struct.h

Parameter must be upper case

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

TD-er committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Update _P036_FrameOLED.ino

for ESP32 the pinMode can not be used, must be set on the hardware page
their the InputPullDown or InputPullUp can be selected

uwekaditz commented:
I tested the 128x32 display on an ESP32. It works fine except the function `pinmode()` can not be used as for an ESP8266.

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -425,7 +425,10 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+#ifndef ESP32
+        // pinMode can not be used for ESP32, must be set on the hardware page
+        pinMode(CONFIG_PIN3, INPUT_PULLUP);     //  Reset pinstate
+#endif
```
Hmm this is new to me.
If this indeed is a limitation on ESP32, please make sure to explain it in the documentation.

uwekaditz replied:
According to the ESP32 datasheet each input can be configured as pulldown (45kOhm), pullup (45kOhm) or high impedance.
With the standard GPIO setting within the plug in I have no chance to set pulldown as I need on my device. Therefore I had to use the hardware page to define the input setting of the GPIO used in the plug in.
In future it would be nice to do it again within the plug in settings.

uwekaditz committed:
[P036] PinMode in Web interface adjustable (even InputPullDown if available)

CHG: PinMode in Web interface adjustable (even InputPullDown if available for ESP32)
CHG: ignore lib HeatpumpIR in platformio_esp82xx_envs for custom_IR_ESP8266_4M1M because linking error (hitting the 32768 characters limitation ???)

TD-er commented on platformio_esp82xx_envs.ini:
```diff
@@ -44,7 +44,7 @@ platform_packages         = ${regular_platform.platform_packages}
                             ${esp8266_4M1M.build_flags} 
                             -DPLUGIN_BUILD_CUSTOM
-lib_ignore                = LittleFS, ESP32_ping, ESP32WebServer, ESP32HTTPUpdateServer, Adafruit GFX Library, LOLIN_EPD, Adafruit ILI9341, Adafruit BusIO, Adafruit NeoPixel, Adafruit Motor Shield V2 Library, RN2xx3 Arduino Library
+lib_ignore                = LittleFS, ESP32_ping, ESP32WebServer, ESP32HTTPUpdateServer, Adafruit GFX Library, LOLIN_EPD, Adafruit ILI9341, Adafruit BusIO, Adafruit NeoPixel, Adafruit Motor Shield V2 Library, RN2xx3 Arduino Library, HeatpumpIR
```
This change makes it impossible to build the HeatpumpIR in the `env:custom_IR_ESP8266_4M1M` env.

uwekaditz replied:
I know, but for this PR the library HeatpumpIR ist not necessary.
I found no other way to avoid the linking error (*.elf -> file not found)

TD-er replied:
Well if it is merged, it is no longer just about this PR.
I will have a look to see what I can do here.

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

TD-er committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz committed:
Merge branch 'mega' into OLED_P036

uwekaditz commented:
I'm done.
From my side everything is ready.

TD-er commented:
OK, will test and have a closer look tomorrow

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -317,6 +355,13 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+      int P036pinmode = getFormItemInt(F("p036_pinmode"));
+      switch (P036pinmode) {
+        case 1: bitWrite(lSettings, 26, true);                                              // Bit 26 Input PullUp
+        case 2: bitWrite(lSettings, 27, true);                                              // Bit 27 Input PullDown
```
Missing `break;` statements here.

uwekaditz replied:
Done, see my next commit.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -425,41 +433,132 @@ void P036_data_struct::display_indicator() {
+    // no footer (indicator) -> returm max. display height
+    return getDisplaySizeSettings(disp_resolution).Height;
+  }
+  else
```
No else statement is needed here.

uwekaditz replied:
Done, see my next commit.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -425,41 +433,132 @@ void P036_data_struct::display_indicator() {
+    // no header
+    return 0;
+  }
+  else
```
No else statement is needed here

uwekaditz replied:
Done, see my next commit.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -425,41 +433,132 @@ void P036_data_struct::display_indicator() {
+    // font found -> calculate top position and space between lines
+    iMaxHeightForFont = FontSizes[iFontIndex].Height * iLinesPerFrame;
+    if (iLinesPerFrame > 1)
+      result.Space = (iHeight-iMaxHeightForFont) / iLinesPerFrame;
```
please add {} here to make clear what line should be depending on this if statement.
Especially since the next line does a `iLinesPerFrame-1`

uwekaditz replied:
Done, see my next commit.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -540,20 +641,20 @@ uint8_t P036_data_struct::display_scroll(ePageScrollSpeed lscrollspeed, int lTas
       }
 
-      if ((PixLengthLineIn > getDisplaySizeSettings(_disp_resolution).Width) && (fScrollTime > 0.0f))
+      if ((PixLengthLineIn > getDisplaySizeSettings(disp_resolution).Width) && (fScrollTime > 0.0f))
```
Comparing floats like this is dangerous.
Also where `fScrollTime` is used is in a divide, so if it is like 0.0000001, then the value of `dPix` may be just any value as it is only an `unit8_t`
So even a value of 0 is possible then.

uwekaditz replied:
Done, see my next commit.
fScrollTime (float) changed to iScrollTime (int),  iScrollTime =scrollTime in ms

TD-er commented on src/src/PluginStructs/P092_data_struct.cpp:
```diff
@@ -60,8 +60,6 @@ void DLBus::AddToErrorLog(const String& string)
 {
   ISR_Receiving = false;
-  IsISRset = true;
-  IsNoData = false;
```
Why is there code for P092 in this PR?

uwekaditz replied:
Done, my mistake.

uwekaditz committed:
Delete P092_data_struct.cpp

uwekaditz committed:
Delete P092_data_struct.h

uwekaditz committed:
Changes requested by TDer

CHG Changes requested by TDer
NEW original P092_data_struct fom mega added

uwekaditz committed:
Update P092_DLbus.ino from mega

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -495,17 +490,21 @@ tFontSettings P036_data_struct::CalculateFontSettings(uint8_t lDefaultLines)
+      // more than one lines per frame -> calculate space inbetween
       result.Space = (iHeight-iMaxHeightForFont) / iLinesPerFrame;
+    }
     result.Top = (iHeight - (iMaxHeightForFont + (result.Space * (iLinesPerFrame-1)))) / 2;
```
```c++
    result.Top = (iHeight - (iMaxHeightForFont + (result.Space * (iLinesPerFrame-1)))) / 2;
```
`iLinesPerFrame` can be 0 here, but it took me a while to realize `result.Space` is 0 at that point.
Maybe you can move the line `result.Space = 0;` to right above `if (iFontIndex >= 0) {` to make it more clear?
It is a prerequisite for this to work, so it should be close or else a reader ("me") will miss it later.

uwekaditz replied:
Done, see my next commit.

TD-er commented on src/src/PluginStructs/P036_data_struct.cpp:
```diff
@@ -645,7 +644,7 @@ uint8_t P036_data_struct::display_scroll(ePageScrollSpeed lscrollspeed, int lTas
       }
 
-      if ((PixLengthLineIn > getDisplaySizeSettings(disp_resolution).Width) && (fScrollTime > 0.0f))
+      if ((PixLengthLineIn > getDisplaySizeSettings(disp_resolution).Width) && (iScrollTime >= 0))
```
should be `iScrollTime > 0` since you are dividing by `iScrollTime` later.

uwekaditz replied:
Thanks, I was a bit off the mark while changing the code.

uwekaditz committed:
Changes requested by TD-er

uwekaditz committed:
Merge branch 'mega' into OLED_P036
